File: apps/services/marketplace-service/src/examples/soft-delete-usage.example.ts
/**
 * Soft Delete Usage Examples
 * أمثلة استخدام الحذف الناعم
 *
 * This file demonstrates how to use the soft delete pattern
 * in the marketplace service.
 */

import { PrismaService } from '../prisma/prisma.service';
import {
  softDelete,
  softDeleteMany,
  restore,
  restoreMany,
  findWithDeleted,
  isDeleted,
  getDeletionMetadata,
} from '@sahool/shared-db';

/**
 * Example Service demonstrating soft delete operations
 */
export class SoftDeleteExamplesService {
  constructor(private readonly prisma: PrismaService) {}

  // ═══════════════════════════════════════════════════════════════════════════
  // Basic Operations - العمليات الأساسية
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Example 1: Soft delete a product
   * مثال 1: حذف ناعم لمنتج
   */
  async softDeleteProduct(productId: string, deletedBy: string) {
    // Method 1: Using helper function
    const product = await softDelete(
      this.prisma.product,
      { id: productId },
      { deletedBy }
    );

    // Method 2: Using Prisma delete (automatically converted to soft delete)
    // This is converted by middleware to an update with deletedAt
    const product2 = await this.prisma.product.delete({
      where: { id: productId },
      // @ts-ignore - deletedBy is not in Prisma types but handled by middleware
      deletedBy,
    });

    return product;
  }

  /**
   * Example 2: Soft delete multiple products
   * مثال 2: حذف ناعم لعدة منتجات
   */
  async softDeleteProductsByCategory(category: string, deletedBy: string) {
    const result = await softDeleteMany(
      this.prisma.product,
      { category },
      { deletedBy }
    );

    console.log(`Deleted ${result.count} products`);
    return result;
  }

  /**
   * Example 3: Find products (automatically excludes deleted ones)
   * مثال 3: البحث عن المنتجات (يستبعد المحذوفة تلقائياً)
   */
  async findActiveProducts() {
    // This automatically excludes soft-deleted products
    const products = await this.prisma.product.findMany({
      where: {
        status: 'AVAILABLE',
      },
    });

    return products;
  }

  /**
   * Example 4: Find products including deleted ones
   * مثال 4: البحث عن المنتجات بما في ذلك المحذوفة
   */
  async findAllProductsIncludingDeleted() {
    // Method 1: Using helper function
    const products = await findWithDeleted(this.prisma.product, {
      where: { status: 'AVAILABLE' },
    });

    // Method 2: Using includeDeleted flag
    const products2 = await this.prisma.product.findMany({
      where: { status: 'AVAILABLE' },
      // @ts-ignore - includeDeleted is handled by middleware
      includeDeleted: true,
    });

    return products;
  }

  /**
   * Example 5: Find only deleted products
   * مثال 5: البحث عن المنتجات المحذوفة فقط
   */
  async findDeletedProducts() {
    const products = await this.prisma.product.findMany({
      where: {
        deletedAt: { not: null },
      },
      // @ts-ignore
      includeDeleted: true,
    });

    return products;
  }

  /**
   * Example 6: Restore a deleted product
   * مثال 6: استعادة منتج محذوف
   */
  async restoreProduct(productId: string) {
    const product = await restore(this.prisma.product, { id: productId });

    console.log(`Product ${productId} restored`);
    return product;
  }

  /**
   * Example 7: Restore multiple products
   * مثال 7: استعادة عدة منتجات
   */
  async restoreProductsByCategory(category: string) {
    const result = await restoreMany(this.prisma.product, { category });

    console.log(`Restored ${result.count} products`);
    return result;
  }

  /**
   * Example 8: Check if a product is deleted
   * مثال 8: التحقق من حذف منتج
   */
  async checkIfProductIsDeleted(productId: string) {
    const product = await this.prisma.product.findUnique({
      where: { id: productId },
      // @ts-ignore
      includeDeleted: true,
    });

    if (!product) {
      return { found: false, deleted: false };
    }

    const deleted = isDeleted(product);
    const metadata = getDeletionMetadata(product);

    return {
      found: true,
      deleted,
      deletionInfo: metadata,
    };
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Advanced Examples - أمثلة متقدمة
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * Example 9: Soft delete with related records
   * مثال 9: حذف ناعم مع السجلات المرتبطة
   */
  async softDeleteOrderWithItems(orderId: string, deletedBy: string) {
    // Get order with items
    const order = await this.prisma.order.findUnique({
      where: { id: orderId },
      include: { items: true },
    });

    if (!order) {
      throw new Error('Order not found');
    }

    // Soft delete the order
    // Note: OrderItem doesn't have soft delete fields, so it won't be deleted
    // If you want to delete items too, add soft delete fields to OrderItem model
    await softDelete(this.prisma.order, { id: orderId }, { deletedBy });

    return { orderId, deletedItemsCount: order.items.length };
  }

  /**
   * Example 10: Cascade soft delete for wallet and related records
   * مثال 10: حذف ناعم متتالي للمحفظة والسجلات المرتبطة
   */
  async softDeleteWalletAndRelated(walletId: string, deletedBy: string) {
    // Soft delete wallet (loans will remain because we might need them for audit)
    await softDelete(this.prisma.wallet, { id: walletId }, { deletedBy });

    // Optionally soft delete related loans
    await softDeleteMany(
      this.prisma.loan,
      { walletId },
      { deletedBy }
    );

    return { walletId, status: 'deleted' };
  }

  /**
   * Example 11: Query with complex filters and soft delete
   * مثال 11: استعلام مع فلاتر معقدة والحذف الناعم
   */
  async findProductsWithComplexFilters() {
    // This query automatically excludes deleted products
    const products = await this.prisma.product.findMany({
      where: {
        AND: [
          { status: 'AVAILABLE' },
          { stock: { gt: 0 } },
          {
            OR: [
              { category: 'SEEDS' },
              { category: 'FERTILIZER' },
            ],
          },
        ],
        // deletedAt: null is automatically added by middleware
      },
      orderBy: {
        createdAt: 'desc',
      },
      take: 10,
    });

    return products;
  }

  /**
   * Example 12: Count active vs deleted records
   * مثال 12: عد السجلات النشطة مقابل المحذوفة
   */
  async getProductStats() {
    // Count active products (excluding deleted)
    const activeCount = await this.prisma.product.count({
      where: { deletedAt: null },
    });

    // Count deleted products
    const deletedCount = await this.prisma.product.count({
      where: { deletedAt: { not: null } },
      // @ts-ignore
      includeDeleted: true,
    });

    // Total count (including deleted)
    const totalCount = await this.prisma.product.count({
      // @ts-ignore
      includeDeleted: true,
    });

    return {
      active: activeCount,
      deleted: deletedCount,
      total: totalCount,
    };
  }

  /**
   * Example 13: Bulk operations with soft delete
   * مثال 13: عمليات جماعية مع الحذف الناعم
   */
  async bulkDeleteExpiredProducts(deletedBy: string) {
    const thirtyDaysAgo = new Date();
    thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);

    const result = await softDeleteMany(
      this.prisma.product,
      {
        status: 'SOLD_OUT',
        updatedAt: { lt: thirtyDaysAgo },
      },
      { deletedBy }
    );

    return result;
  }

  /**
   * Example 14: Restore with validation
   * مثال 14: استعادة مع التحقق
   */
  async restoreProductWithValidation(productId: string) {
    // First, find the product including deleted ones
    const product = await this.prisma.product.findUnique({
      where: { id: productId },
      // @ts-ignore
      includeDeleted: true,
    });

    if (!product) {
      throw new Error('Product not found');
    }

    if (!isDeleted(product)) {
      throw new Error('Product is not deleted');
    }

    // Check deletion metadata
    const metadata = getDeletionMetadata(product);
    console.log(`Product was deleted by: ${metadata?.deletedBy}`);
    console.log(`Deleted at: ${metadata?.deletedAt}`);

    // Restore the product
    const restored = await restore(this.prisma.product, { id: productId });

    return restored;
  }

  /**
   * Example 15: Get deletion audit trail
   * مثال 15: الحصول على سجل التدقيق للحذف
   */
  async getDeletionAuditTrail(startDate: Date, endDate: Date) {
    // Find all products deleted in the date range
    const deletedProducts = await this.prisma.product.findMany({
      where: {
        deletedAt: {
          gte: startDate,
          lte: endDate,
        },
      },
      // @ts-ignore
      includeDeleted: true,
      select: {
        id: true,
        name: true,
        deletedAt: true,
        deletedBy: true,
        category: true,
      },
    });

    return deletedProducts.map(product => ({
      productId: product.id,
      productName: product.name,
      category: product.category,
      ...getDeletionMetadata(product),
    }));
  }
}

File: apps/services/billing-core/examples/soft_delete_usage_example.py
"""
Soft Delete Usage Examples for Billing Core
============================================

This module demonstrates how to use the soft delete pattern with SQLAlchemy
in the billing service.

Author: SAHOOL Team
"""

# Import the soft delete utilities
import sys
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

sys.path.append("/home/user/sahool-unified-v15-idp/packages/shared-db/src")
from soft_delete_sqlalchemy import (
    count_active_records,
    count_deleted_records,
    get_active_records,
    get_deleted_records,
    get_deletion_metadata,
    restore_many,
    restore_record,
    soft_delete_many,
    soft_delete_record,
)

# Import your models (these would be your actual billing models)
from src.models import Invoice, Payment, Subscription


class SoftDeleteExamplesService:
    """
    Service class demonstrating soft delete operations in billing core.
    """

    def __init__(self, session: Session):
        self.session = session

    # ════════════════════════════════════════════════════════════════════════
    # Basic Operations - العمليات الأساسية
    # ════════════════════════════════════════════════════════════════════════

    def soft_delete_subscription(
        self, subscription_id: str, deleted_by: str
    ) -> Subscription | None:
        """
        Example 1: Soft delete a subscription
        مثال 1: حذف ناعم لاشتراك

        Args:
            subscription_id: ID of the subscription to delete
            deleted_by: User ID performing the deletion

        Returns:
            The soft-deleted subscription or None if not found
        """
        subscription = soft_delete_record(
            self.session, Subscription, subscription_id, deleted_by=deleted_by
        )

        if subscription:
            print(f"Subscription {subscription_id} soft-deleted by {deleted_by}")

        self.session.commit()
        return subscription

    def get_active_subscriptions(
        self, tenant_id: str | None = None
    ) -> list[Subscription]:
        """
        Example 2: Get all active subscriptions
        مثال 2: الحصول على جميع الاشتراكات النشطة

        Args:
            tenant_id: Optional tenant ID to filter by

        Returns:
            List of active subscriptions
        """
        filters = {}
        if tenant_id:
            filters["tenant_id"] = tenant_id

        subscriptions = get_active_records(self.session, Subscription, **filters)

        print(f"Found {len(subscriptions)} active subscriptions")
        return subscriptions

    def get_deleted_subscriptions(self) -> list[Subscription]:
        """
        Example 3: Get all deleted subscriptions
        مثال 3: الحصول على جميع الاشتراكات المحذوفة

        Returns:
            List of soft-deleted subscriptions
        """
        subscriptions = get_deleted_records(self.session, Subscription)

        print(f"Found {len(subscriptions)} deleted subscriptions")
        return subscriptions

    def restore_subscription(self, subscription_id: str) -> Subscription | None:
        """
        Example 4: Restore a soft-deleted subscription
        مثال 4: استعادة اشتراك محذوف

        Args:
            subscription_id: ID of the subscription to restore

        Returns:
            The restored subscription or None if not found
        """
        subscription = restore_record(self.session, Subscription, subscription_id)

        if subscription:
            print(f"Subscription {subscription_id} restored successfully")

        self.session.commit()
        return subscription

    # ════════════════════════════════════════════════════════════════════════
    # Advanced Operations - عمليات متقدمة
    # ════════════════════════════════════════════════════════════════════════

    def soft_delete_expired_trials(self, deleted_by: str) -> int:
        """
        Example 5: Soft delete all expired trial subscriptions
        مثال 5: حذف ناعم لجميع الاشتراكات التجريبية المنتهية

        Args:
            deleted_by: User ID performing the deletion

        Returns:
            Number of subscriptions deleted
        """
        # Find expired trials
        today = datetime.utcnow().date()

        # First, get the subscriptions to delete
        expired_trials = (
            self.session.query(Subscription)
            .filter(
                Subscription.status == "trial",
                Subscription.trial_end_date < today,
                Subscription.deleted_at.is_(None),
            )
            .all()
        )

        # Soft delete each one
        for subscription in expired_trials:
            subscription.soft_delete(deleted_by=deleted_by)

        count = len(expired_trials)
        self.session.commit()

        print(f"Soft-deleted {count} expired trial subscriptions")
        return count

    def cascade_delete_tenant_data(self, tenant_id: str, deleted_by: str) -> dict:
        """
        Example 6: Cascade soft delete all data for a tenant
        مثال 6: حذف ناعم متتالي لجميع بيانات مستأجر

        Args:
            tenant_id: Tenant ID
            deleted_by: User ID performing the deletion

        Returns:
            Dictionary with counts of deleted records
        """
        counts = {}

        # Delete subscriptions
        counts["subscriptions"] = soft_delete_many(
            self.session, Subscription, deleted_by=deleted_by, tenant_id=tenant_id
        )

        # Delete invoices
        counts["invoices"] = soft_delete_many(
            self.session, Invoice, deleted_by=deleted_by, tenant_id=tenant_id
        )

        # Delete payments
        counts["payments"] = soft_delete_many(
            self.session, Payment, deleted_by=deleted_by, tenant_id=tenant_id
        )

        self.session.commit()

        print(f"Cascade deleted data for tenant {tenant_id}")
        print(f"  - Subscriptions: {counts['subscriptions']}")
        print(f"  - Invoices: {counts['invoices']}")
        print(f"  - Payments: {counts['payments']}")

        return counts

    def get_subscription_stats(self) -> dict:
        """
        Example 7: Get subscription statistics
        مثال 7: الحصول على إحصائيات الاشتراكات

        Returns:
            Dictionary with subscription counts
        """
        active_count = count_active_records(self.session, Subscription)
        deleted_count = count_deleted_records(self.session, Subscription)
        total_count = active_count + deleted_count

        stats = {
            "active": active_count,
            "deleted": deleted_count,
            "total": total_count,
            "deletion_rate": (
                (deleted_count / total_count * 100) if total_count > 0 else 0
            ),
        }

        print("Subscription Statistics:")
        print(f"  Active: {stats['active']}")
        print(f"  Deleted: {stats['deleted']}")
        print(f"  Total: {stats['total']}")
        print(f"  Deletion Rate: {stats['deletion_rate']:.2f}%")

        return stats

    def check_subscription_deletion_status(self, subscription_id: str) -> dict:
        """
        Example 8: Check if a subscription is deleted and get metadata
        مثال 8: التحقق من حذف اشتراك والحصول على البيانات الوصفية

        Args:
            subscription_id: Subscription ID to check

        Returns:
            Dictionary with deletion status and metadata
        """
        # Get subscription including deleted ones
        subscription = (
            self.session.query(Subscription)
            .filter(Subscription.id == subscription_id)
            .first()
        )

        if not subscription:
            return {"found": False, "deleted": False, "metadata": None}

        is_deleted = subscription.is_deleted()
        metadata = get_deletion_metadata(subscription)

        result = {"found": True, "deleted": is_deleted, "metadata": metadata}

        if is_deleted:
            print(f"Subscription {subscription_id} is deleted")
            print(f"  Deleted at: {metadata['deleted_at']}")
            print(f"  Deleted by: {metadata['deleted_by']}")
        else:
            print(f"Subscription {subscription_id} is active")

        return result

    def get_deletion_audit_trail(
        self, start_date: datetime, end_date: datetime
    ) -> list[dict]:
        """
        Example 9: Get deletion audit trail for a date range
        مثال 9: الحصول على سجل تدقيق الحذف لنطاق تاريخ

        Args:
            start_date: Start of date range
            end_date: End of date range

        Returns:
            List of deletion records
        """
        # Find all subscriptions deleted in the date range
        deleted_subscriptions = (
            self.session.query(Subscription)
            .filter(
                Subscription.deleted_at >= start_date,
                Subscription.deleted_at <= end_date,
            )
            .all()
        )

        audit_trail = []
        for subscription in deleted_subscriptions:
            metadata = get_deletion_metadata(subscription)
            audit_trail.append(
                {
                    "subscription_id": str(subscription.id),
                    "tenant_id": subscription.tenant_id,
                    "plan_id": subscription.plan_id,
                    "deleted_at": metadata["deleted_at"],
                    "deleted_by": metadata["deleted_by"],
                }
            )

        print(f"Found {len(audit_trail)} deletions between {start_date} and {end_date}")
        return audit_trail

    def restore_recent_deletions(self, hours: int = 24) -> int:
        """
        Example 10: Restore all subscriptions deleted in the last N hours
        مثال 10: استعادة جميع الاشتراكات المحذوفة في آخر N ساعة

        Args:
            hours: Number of hours to look back

        Returns:
            Number of subscriptions restored
        """
        cutoff_time = datetime.utcnow() - timedelta(hours=hours)

        # Find recently deleted subscriptions
        recent_deleted = (
            self.session.query(Subscription)
            .filter(Subscription.deleted_at >= cutoff_time)
            .all()
        )

        count = 0
        for subscription in recent_deleted:
            subscription.restore()
            count += 1

        self.session.commit()

        print(f"Restored {count} subscriptions deleted in the last {hours} hours")
        return count

    # ════════════════════════════════════════════════════════════════════════
    # Query Examples - أمثلة الاستعلامات
    # ════════════════════════════════════════════════════════════════════════

    def find_active_subscriptions_with_filters(
        self, plan_id: str | None = None, status: str | None = None
    ) -> list[Subscription]:
        """
        Example 11: Complex query with soft delete filtering
        مثال 11: استعلام معقد مع تصفية الحذف الناعم

        Args:
            plan_id: Optional plan ID to filter by
            status: Optional status to filter by

        Returns:
            List of matching active subscriptions
        """
        # Build query
        query = self.session.query(Subscription)

        # Add soft delete filter
        query = Subscription.filter_active(query)

        # Add optional filters
        if plan_id:
            query = query.filter(Subscription.plan_id == plan_id)

        if status:
            query = query.filter(Subscription.status == status)

        # Order by created date
        query = query.order_by(Subscription.created_at.desc())

        subscriptions = query.all()

        print(f"Found {len(subscriptions)} active subscriptions matching filters")
        return subscriptions

    def bulk_restore_by_tenant(self, tenant_id: str) -> int:
        """
        Example 12: Bulk restore all deleted subscriptions for a tenant
        مثال 12: استعادة جماعية لجميع الاشتراكات المحذوفة لمستأجر

        Args:
            tenant_id: Tenant ID

        Returns:
            Number of subscriptions restored
        """
        count = restore_many(self.session, Subscription, tenant_id=tenant_id)

        self.session.commit()

        print(f"Restored {count} subscriptions for tenant {tenant_id}")
        return count

    def cleanup_old_deleted_records(self, days: int = 180) -> int:
        """
        Example 13: Permanently delete old soft-deleted records
        مثال 13: حذف دائم للسجلات المحذوفة القديمة

        CAUTION: This permanently deletes records. Use with care!

        Args:
            days: Number of days after which to permanently delete

        Returns:
            Number of records permanently deleted
        """
        cutoff_date = datetime.utcnow() - timedelta(days=days)

        # Find old deleted subscriptions
        old_deleted = (
            self.session.query(Subscription)
            .filter(Subscription.deleted_at < cutoff_date)
            .all()
        )

        count = len(old_deleted)

        # Permanently delete them
        for subscription in old_deleted:
            self.session.delete(subscription)  # Hard delete

        self.session.commit()

        print(f"Permanently deleted {count} subscriptions older than {days} days")
        return count


# ════════════════════════════════════════════════════════════════════════════
# Usage Example
# ════════════════════════════════════════════════════════════════════════════


def main():
    """
    Main function demonstrating usage of the soft delete examples.
    """
    from src.database import SessionLocal

    # Create a database session
    session = SessionLocal()

    try:
        # Create the examples service
        SoftDeleteExamplesService(session)

        # Example 1: Soft delete a subscription
        # service.soft_delete_subscription('sub-123', deleted_by='admin-456')

        # Example 2: Get active subscriptions
        # active_subs = service.get_active_subscriptions(tenant_id='tenant-789')

        # Example 3: Get subscription stats
        # stats = service.get_subscription_stats()

        # Example 4: Restore a subscription
        # service.restore_subscription('sub-123')

        # Example 5: Get deletion audit trail
        # from datetime import datetime, timedelta
        # start = datetime.utcnow() - timedelta(days=7)
        # end = datetime.utcnow()
        # audit_trail = service.get_deletion_audit_trail(start, end)

        print("Soft delete examples completed successfully!")

    finally:
        session.close()


if __name__ == "__main__":
    main()
